ocaml/xcp-rrdd/bin/rrdd/rrdd_server.ml

```diff
@@ -715,9 +715,25 @@ module Plugin = struct
          specified unique ID. If the plugin is not registered, -1 is returned. *)
       let next_reading (uid : P.uid) : float =
         let open Rrdd_shared in
+        let ( --- ) a b = Mtime.Span.abs_diff a b in
         if with_lock registered_m (fun _ -> Hashtbl.mem registered uid) then
-          with_lock last_loop_end_time_m (fun _ ->
-              !last_loop_end_time +. !timeslice -. Unix.gettimeofday ()
+          with_lock next_iteration_start_m (fun _ ->
+              let current = Mtime_clock.count from_loop_start in
+              let diff =
+                Mtime.Span.to_float_ns
+                  (!Rrdd_shared.next_iteration_start
+                  --- Mtime_clock.count from_loop_start
+                  )
+                /. 1_000_000_000.
+              in
+              if
+                Mtime.Span.is_longer
+                  ~than:!Rrdd_shared.next_iteration_start
+                  current
+              then
+                diff *. -1.
```

Comment:
Hmmm, this looks like it should be done with a [Timer](https://github.com/xapi-project/xen-api/blob/master/ocaml/libs/clock/timer.mli#L8):
- you have the case where there's enough time to wait
- you want to know by how much you missed a deadline

So instead of a counter from the start of the loop, you set up a timer with 5 seconds of duration, and on each iteration you match on `Timer.remaining next_start` and find out whether the deadline was missed or not, and by how much. After that you can add 5 seconds to the timer and ask how much time there is left to to wait. (by using `extend_by` and `remaining` again)